## Assignment 2: Linear Discriminant Functions and Support Vector Machines

### Part 1: Multi-class support vector classifier with dot-product kernel and 1-norm soft margin using the MNIST training data set

In [1]:
from mnist import MNIST
import numpy as np
import sklearn as sk
from sklearn import model_selection, metrics, svm
import matplotlib.pyplot as plt
mnist = MNIST('./dataset')

In [2]:
#Load training and testing data from the dataset
x_train, y_train = mnist.load_training()
x_test, y_test = mnist.load_testing()

#assign training images to X_train and training labels to y_train
x_train = np.asarray(x_train).astype(np.float32)/255.0
y_train = np.asarray(y_train).astype(np.int32)

#assign testing images to X_test and testing labels to y_test
x_test = np.asarray(x_test).astype(np.float32)/255.0
y_test = np.asarray(y_test).astype(np.int32)

X_train = x_train.reshape(-1,784)
Y_train = y_train.reshape(X_train.shape[0],)
X_test = x_test.reshape(-1,784)
Y_test = y_test.reshape(X_test.shape[0],)

Here, C is a regularization parameter that controls the trade-off between maximizing the margin and minimizing the training error. Small C tends to emphasize the margin while ignoring the outliers in the training data, while large C may tend to overfit the training data. We calculate the accuracy of the Support Vector Classifier using k-fold Cross-Validation by combining the test data into the training data and taking the value of K=7.

In [16]:
#Using a C value of 1
SVM = svm.LinearSVC(C=1, penalty='l1', dual=False)

cv_score = model_selection.cross_validate(SVM, X=np.concatenate((X_train, X_test)), y=np.concatenate((Y_train, Y_test)), cv=7)

accuracy_C_1 = np.mean(cv_score['test_score'])
accuracy_C_1

0.9140288411646743

In [11]:
#Using a C value of 2
SVM = svm.LinearSVC(C=2, penalty='l1', dual=False)

cv_score = model_selection.cross_validate(SVM, X=np.concatenate((X_train, X_test)), y=np.concatenate((Y_train, Y_test)), cv=7)
accuracy_C_2 = np.mean(cv_score['test_score'])
accuracy_C_2

0.9135859925632452

In [12]:
#Using a C value of 5
SVM = svm.LinearSVC(C=5, penalty='l1', dual=False)

cv_score = model_selection.cross_validate(SVM, X=np.concatenate((X_train, X_test)), y=np.concatenate((Y_train, Y_test)), cv=7)
accuracy_C_5 = np.mean(cv_score['test_score'])
accuracy_C_5

0.9129574625268236

In [13]:
#Using a C value of 0.1
SVM = svm.LinearSVC(C=0.1, penalty='l1', dual=False)

cv_score = model_selection.cross_validate(SVM, X=np.concatenate((X_train, X_test)), y=np.concatenate((Y_train, Y_test)), cv=7)
accuracy_C_0_1 = np.mean(cv_score['test_score'])
accuracy_C_0_1

0.9138574083251046

In [9]:
#Using a C value of 0.5
SVM = svm.LinearSVC(C=0.5, penalty='l1', dual=False)
cv_score = model_selection.cross_validate(SVM, X=np.concatenate((X_train, X_test)), y=np.concatenate((Y_train, Y_test)), cv=7)
accuracy_C_0_5 = np.mean(cv_score['test_score'])
accuracy_C_0_5

0.9141860025839611

In [15]:
#Using a C value of 0.6
SVM = svm.LinearSVC(C=0.6, penalty='l1', dual=False)

cv_score = model_selection.cross_validate(SVM, X=np.concatenate((X_train, X_test)), y=np.concatenate((Y_train, Y_test)), cv=7)
accuracy_C_0_6 = np.mean(cv_score['test_score'])
accuracy_C_0_6

0.9142860183029609

In [14]:
#Using a C value of 0.7
SVM = svm.LinearSVC(C=0.7, penalty='l1', dual=False)

cv_score = model_selection.cross_validate(SVM, X=np.concatenate((X_train, X_test)), y=np.concatenate((Y_train, Y_test)), cv=7)
accuracy_C_0_7 = np.mean(cv_score['test_score'])
accuracy_C_0_7

0.9141431368748157

From the above accuracy values determined using Cross Validation, the value of hyperparameter C=0.6 gives the best accuracy of 91.42%.

### Part 2: Lagrange Primal and Dual Problems

Given features $(x_{1}, y_{1}), ... ,(x_{N}, y_{N})$, where    $y_{1},..,y_{N} \epsilon$ {-1,1}

The hard margin for Primal problem is

Minimize $w^{T}.w + C\sum_{i=1}^{N}\xi_{i}$ where w is the separating vector, $w_{T}.w$ is the dot product, and $\xi_{i}$ is the error made by the separating vector w on feature $(x_{i}, y_{i})$

Subject to $y_{i}.(w^{T}.x_{i})\geq 1 - \xi_{i}$ and $\xi_{i}\geq0$ for i = 1,..,N

Note : we explicitly add the bias to the constraint, rather than treating it as a part of $x_{i}$, to aide in calculation.

The corresponding Lagrangian for the 1-norm soft margin optimization problem is,

$L(w, b, \xi, \alpha, r) = \frac{1}{2}w^{T}.w + C\sum_{i=1}^{N}\xi_{i} - \sum_{i=1}^{N}\alpha_{i}[y_{i}(w^{T}.x_{i}+b)-1+\xi_{i}] - \sum_{i=1}^{N}r_{i}\xi_{i}$


where $\alpha_{i}$ and $r_{i}$  are Lagrangian multipliers and 
$\alpha_{i}\geq0$ and $r_{i}\geq0 $


Differentiating the Legrangian with respect to w, $\xi$ and b, we have

$\frac{\partial L(w,b,\xi,\alpha,r)}{\partial w} = w - \sum_{i=1}^{N}y_{i}\alpha_{i}x_{i}=0$

$w = \sum_{i=1}^{N}y_{i}\alpha_{i}x_{i}$

$\frac{\partial L(w,b,\xi,\alpha,r)}{\partial \xi_{i}} = C-\alpha_{i} - r_{i} = 0$

$ C = \alpha_{i} + r_{i}$


$\frac{\partial L(w,b,\xi,\alpha,r)}{\partial b} = \sum_{i=1}^{N}y_{i}\alpha_{i} = 0$

Resubstituting these relations into the primal, we have the following dual problem:


maximize $ L_{d}(\alpha, r) = \sum_{i=1}^{N}\alpha_{i} - \frac{1}{2}\sum_{i=1}^{N}\sum_{j=1}^{N}y_{i}y_{j}\alpha_{i}\alpha_{j}(x_{j}^{T}.x_{i}) - \sum_{i=1}^{N}\alpha_{i}\xi_{i} - \sum_{i=1}^{N}r_{i}\xi_{i} - C\sum_{i=1}^{N}\xi_{i}$

    
 $= \sum_{i=1}^{N}\alpha_{i}- \frac{1}{2}\sum_{i=1}^{N}\sum_{j=1}^{N}y_{i}y_{j}\alpha_{i}\alpha_{j}(x_{j}^{T}.x_{i})$


Subjected to the constraint    $0\leq\alpha_{i}\leq C$,   $\sum_{i=1}^{N}\alpha_{i}y_{i} = 0$

Given the relations,

$ C = \alpha_{i} + r_{i}$
and since $\alpha_{i}\geq0 $ and $r_{i}\geq0$, we have $0\leq\alpha_{i}\leq C$


The objective function of the dual problem is identical to the linear seperable problem. Solving this problem we get optimal decision plane w and b with the margin given as,

$(\sum_{i=1}^{sv}\sum_{j=1}^{sv}\alpha_{i}\alpha_{j}y_{i}y_{j}(x_{i}^{T}.x_{j}))^{\frac{-1}{2}}$

where sv is the set of support vectors - data points that lie along the margin, that is, the data points closest to the hyperplane. 

From the equation,
$y_{i}.(w^{T}.x_{i})\geq 1 - \xi_{i}$ and $\xi_{i}\geq0$ for i = 1,..,N

the support vectors are those $x_i$s where the corresponding $\xi_{i}$>0 In other words, they are the data points that are either misclassified, or close to the boundary.


Also, using the relation, 

$w = \sum_{i=1}^{N}y_{i}\alpha_{i}x_{i}$ 

and substituting into the equation,

$  w^{T}x + b = (\sum_{i=1}^{N}\alpha_{i}y_{i}x_{i})^{T}.x +b $

$= \sum_{i=1}^{N}\alpha_{i}y_{i}<x_{i}.x> + b$

Hence, if we’ve found the $\alpha_{i}s$, in order to make a prediction, we have to calculate a quantity that depends only on the inner product between x and the support vectors.

The advantage of maximizing the margin is to make it easier to separate data points that belong to the two different classes. 

The benefits of solving the dual problem over the primal problem are:

1) It lends itself easily to the Kernel Trick. The optimization problem can be written as:

$max_{\alpha} = \sum_{i=1}^{N}\alpha_{i} - \frac{1}{2}\sum_{i=1}^{N}\sum_{j=1}^{N}y_{i}y_{j}\alpha_{i}\alpha_{j}K(x_{i}x_{j})$

such that $\forall \alpha_{i} \geq 0$ and $\sum_{i=1}^{N}y_{i}\alpha_{i} = 0$


This is almost the same as the original dual formulation, except we compute the kernel function instead of ordinary dot product. This is not possible in the primal formulation where it would be necessary the mapping for each data point.

2) Classifying a new data point is much easier in dual as the number of parameters is only equal to the number of data points.

## References:

1) https://scikit-learn.org/stable/modules/generated/sklearn.svm.LinearSVC.html?highlight=svc#sklearn.svm.LinearSVC


2) http://fourier.eng.hmc.edu/e161/lectures/svm/node7.html


3) http://cs229.stanford.edu/notes/cs229-notes3.pdf


4) https://people.eecs.berkeley.edu/~jordan/courses/281B-spring04/lectures/lec6.pdf


5) https://www.quora.com/Why-is-solving-in-the-dual-easier-than-solving-in-the-primal-What-advantages-do-we-get-from-solving-in-the-dual
